File: javascript/all-your-base/all-your-base.js
//
// This is only a SKELETON file for the 'All Your Base' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const convert = (digits, inputBase, outputBase) => {
  if (inputBase <= 1) {
    throw new Error('Wrong input base');
  } else if (outputBase <= 1) {
    throw new Error('Wrong output base');
  };
  if (digits.length === 0) {
    throw new Error('Input has wrong format');
  } else if (digits[0] === 0) {
    if (digits.length === 1) {
      return [0];
    } else {
      throw new Error('Input has wrong format');
    };
  };
  let number = 0;
  for (let i = 0; i < digits.length; i++) {
    if (digits[digits.length-i-1] < 0 || digits[digits.length-i-1] >= inputBase) {
      throw new Error('Input has wrong format');
    };
    number += digits[digits.length-i-1] * (inputBase ** i);
  };
  let output = [];
  while (number > 0) {
    output.unshift(number % outputBase);
    number = (number - (number % outputBase)) / outputBase;
  };
  return output
};

File: elixir/all-your-base/lib/all_your_base.ex
defmodule AllYourBase do
  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """

  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(digits, input_base, output_base) do
    cond do
      output_base < 2 -> {:error, "output base must be >= 2"}
      input_base < 2 -> {:error, "input base must be >= 2"}
      Enum.any?(digits, &(&1 < 0 || &1 >= input_base)) -> {:error, "all digits must be >= 0 and < input base"}
      true -> {:ok, find_digits(base_ten(digits, input_base, 0), output_base, [])}
    end
  end

  defp base_ten([], _, prev), do: prev
  defp base_ten(_digits = [head | tail], input_base, prev) do
    base_ten(tail, input_base, input_base * prev + head)
  end

  defp find_digits(0, _, []), do: [0]
  defp find_digits(0, _, output_digits), do: output_digits
  defp find_digits(number, output_base, output_digits) do
    find_digits(div(number, output_base), output_base, [rem(number, output_base) | output_digits]) 
  end
end
